File: backend/security.py
import re
import hashlib
import secrets
import string
from typing import Optional, Dict, Any
from werkzeug.security import generate_password_hash, check_password_hash
import jwt
from datetime import datetime, timedelta
from config import Config

class SecurityUtils:
    """Security utilities for input validation, sanitization, and authentication"""
    
    @staticmethod
    def validate_email(email: str) -> bool:
        """Validate email format"""
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return bool(re.match(pattern, email))
    
    @staticmethod
    def validate_password(password: str) -> Dict[str, Any]:
        """Validate password strength"""
        errors = []
        
        if len(password) < Config.PASSWORD_MIN_LENGTH:
            errors.append(f"Password must be at least {Config.PASSWORD_MIN_LENGTH} characters long")
        
        if Config.PASSWORD_REQUIRE_UPPERCASE and not re.search(r'[A-Z]', password):
            errors.append("Password must contain at least one uppercase letter")
        
        if Config.PASSWORD_REQUIRE_LOWERCASE and not re.search(r'[a-z]', password):
            errors.append("Password must contain at least one lowercase letter")
        
        if Config.PASSWORD_REQUIRE_DIGITS and not re.search(r'\d', password):
            errors.append("Password must contain at least one digit")
        
        if Config.PASSWORD_REQUIRE_SPECIAL and not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
            errors.append("Password must contain at least one special character")
        
        return {
            'valid': len(errors) == 0,
            'errors': errors
        }
    
    @staticmethod
    def sanitize_input(input_str: str) -> str:
        """Sanitize user input to prevent XSS"""
        if not input_str:
            return ""
        
        # Remove potentially dangerous characters
        dangerous_chars = ['<', '>', '"', "'", '&', 'javascript:', 'vbscript:', 'onload', 'onerror']
        sanitized = input_str
        
        for char in dangerous_chars:
            sanitized = sanitized.replace(char, '')
        
        # Remove script tags
        sanitized = re.sub(r'<script[^>]*>.*?</script>', '', sanitized, flags=re.IGNORECASE | re.DOTALL)
        
        return sanitized.strip()
    
    @staticmethod
    def generate_secure_token() -> str:
        """Generate a secure random token"""
        return secrets.token_urlsafe(32)
    
    @staticmethod
    def hash_password(password: str) -> str:
        """Hash password securely"""
        return generate_password_hash(password, method='pbkdf2:sha256')
    
    @staticmethod
    def verify_password(password: str, hashed_password: str) -> bool:
        """Verify password against hash"""
        return check_password_hash(hashed_password, password)
    
    @staticmethod
    def generate_jwt_token(user_id: str, user_type: str) -> str:
        """Generate JWT token for user"""
        payload = {
            'user_id': user_id,
            'user_type': user_type,
            'exp': datetime.utcnow() + Config.JWT_ACCESS_TOKEN_EXPIRES,
            'iat': datetime.utcnow()
        }
        return jwt.encode(payload, Config.JWT_SECRET_KEY, algorithm='HS256')
    
    @staticmethod
    def verify_jwt_token(token: str) -> Optional[Dict[str, Any]]:
        """Verify and decode JWT token"""
        try:
            payload = jwt.decode(token, Config.JWT_SECRET_KEY, algorithms=['HS256'])
            return payload
        except jwt.ExpiredSignatureError:
            return None
        except jwt.InvalidTokenError:
            return None
    
    @staticmethod
    def validate_file_extension(filename: str) -> bool:
        """Validate file extension"""
        if '.' not in filename:
            return False
        ext = filename.rsplit('.', 1)[1].lower()
        return ext in Config.ALLOWED_EXTENSIONS
    
    @staticmethod
    def validate_file_size(file_size: int) -> bool:
        """Validate file size"""
        return file_size <= Config.MAX_CONTENT_LENGTH
    
    @staticmethod
    def generate_csrf_token() -> str:
        """Generate CSRF token"""
        return secrets.token_hex(32)
    
    @staticmethod
    def validate_csrf_token(token: str, stored_token: str) -> bool:
        """Validate CSRF token"""
        return secrets.compare_digest(token, stored_token)
    
    @staticmethod
    def sanitize_filename(filename: str) -> str:
        """Sanitize filename for safe storage"""
        # Remove dangerous characters
        filename = re.sub(r'[<>:"/\\|?*]', '', filename)
        # Limit length
        if len(filename) > 255:
            name, ext = filename.rsplit('.', 1)
            filename = name[:255-len(ext)-1] + '.' + ext
        return filename
    
    @staticmethod
    def validate_phone_number(phone: str) -> bool:
        """Validate phone number format"""
        # Remove all non-digit characters
        digits_only = re.sub(r'\D', '', phone)
        # Check if it's a valid length (7-15 digits)
        return 7 <= len(digits_only) <= 15
    
    @staticmethod
    def validate_license_number(license_number: str) -> bool:
        """Validate license number format"""
        # Remove spaces and convert to uppercase
        license_number = re.sub(r'\s+', '', license_number).upper()
        # Check if it contains only alphanumeric characters and common separators
        return bool(re.match(r'^[A-Z0-9\-\.\/]+$', license_number))
    
    @staticmethod
    def log_security_event(event_type: str, user_id: Optional[str] = None, details: Optional[str] = None):
        """Log security events"""
        timestamp = datetime.utcnow().isoformat()
        log_entry = f"[SECURITY] {timestamp} - {event_type}"
        if user_id:
            log_entry += f" - User: {user_id}"
        if details:
            log_entry += f" - Details: {details}"
        
        # In production, this should go to a proper logging system
        print(log_entry) 

File: frontend/js/security.js
/**
 * Frontend Security Utilities
 * Provides input validation, sanitization, and security functions
 */

class SecurityUtils {
    /**
     * Sanitize user input to prevent XSS attacks
     * @param {string} input - The input to sanitize
     * @returns {string} - Sanitized input
     */
    static sanitizeInput(input) {
        if (typeof input !== 'string') {
            return '';
        }
        
        // Remove potentially dangerous characters and scripts
        return input
            .replace(/[<>\"'&]/g, '')
            .replace(/javascript:/gi, '')
            .replace(/vbscript:/gi, '')
            .replace(/on\w+\s*=/gi, '')
            .replace(/<script[^>]*>.*?<\/script>/gi, '')
            .trim();
    }
    
    /**
     * Validate email format
     * @param {string} email - Email to validate
     * @returns {boolean} - True if valid email
     */
    static validateEmail(email) {
        const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
        return emailRegex.test(email);
    }
    
    /**
     * Validate phone number format
     * @param {string} phone - Phone number to validate
     * @returns {boolean} - True if valid phone number
     */
    static validatePhone(phone) {
        const phoneRegex = /^[\d\s\-\+\(\)]{7,15}$/;
        return phoneRegex.test(phone);
    }
    
    /**
     * Validate password strength
     * @param {string} password - Password to validate
     * @returns {object} - Validation result with errors array
     */
    static validatePassword(password) {
        const errors = [];
        
        if (!password || password.length < 8) {
            errors.push('Password must be at least 8 characters long');
        }
        
        if (!/[A-Z]/.test(password)) {
            errors.push('Password must contain at least one uppercase letter');
        }
        
        if (!/[a-z]/.test(password)) {
            errors.push('Password must contain at least one lowercase letter');
        }
        
        if (!/\d/.test(password)) {
            errors.push('Password must contain at least one digit');
        }
        
        if (!/[!@#$%^&*(),.?":{}|<>]/.test(password)) {
            errors.push('Password must contain at least one special character');
        }
        
        return {
            valid: errors.length === 0,
            errors: errors
        };
    }
    
    /**
     * Validate license number format
     * @param {string} licenseNumber - License number to validate
     * @returns {boolean} - True if valid license number
     */
    static validateLicenseNumber(licenseNumber) {
        const licenseRegex = /^[A-Z0-9\-\.\/\s]+$/;
        return licenseRegex.test(licenseNumber);
    }
    
    /**
     * Validate file type
     * @param {File} file - File to validate
     * @param {Array} allowedTypes - Array of allowed MIME types
     * @returns {boolean} - True if valid file type
     */
    static validateFileType(file, allowedTypes = ['image/jpeg', 'image/png', 'image/gif', 'application/pdf']) {
        return allowedTypes.includes(file.type);
    }
    
    /**
     * Validate file size
     * @param {File} file - File to validate
     * @param {number} maxSize - Maximum size in bytes (default: 16MB)
     * @returns {boolean} - True if file size is acceptable
     */
    static validateFileSize(file, maxSize = 16 * 1024 * 1024) {
        return file.size <= maxSize;
    }
    
    /**
     * Sanitize filename for safe storage
     * @param {string} filename - Filename to sanitize
     * @returns {string} - Sanitized filename
     */
    static sanitizeFilename(filename) {
        return filename
            .replace(/[<>:"/\\|?*]/g, '')
            .substring(0, 255);
    }
    
    /**
     * Generate CSRF token
     * @returns {string} - CSRF token
     */
    static generateCSRFToken() {
        return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
    }
    
    /**
     * Validate form data
     * @param {object} formData - Form data to validate
     * @param {object} rules - Validation rules
     * @returns {object} - Validation result
     */
    static validateForm(formData, rules) {
        const errors = {};
        
        for (const [field, rule] of Object.entries(rules)) {
            const value = formData[field];
            
            if (rule.required && (!value || value.trim() === '')) {
                errors[field] = `${field} is required`;
                continue;
            }
            
            if (value && rule.type === 'email' && !this.validateEmail(value)) {
                errors[field] = 'Invalid email format';
            }
            
            if (value && rule.type === 'phone' && !this.validatePhone(value)) {
                errors[field] = 'Invalid phone number format';
            }
            
            if (value && rule.type === 'password') {
                const passwordValidation = this.validatePassword(value);
                if (!passwordValidation.valid) {
                    errors[field] = passwordValidation.errors.join(', ');
                }
            }
            
            if (value && rule.minLength && value.length < rule.minLength) {
                errors[field] = `${field} must be at least ${rule.minLength} characters long`;
            }
            
            if (value && rule.maxLength && value.length > rule.maxLength) {
                errors[field] = `${field} must be no more than ${rule.maxLength} characters long`;
            }
        }
        
        return {
            valid: Object.keys(errors).length === 0,
            errors: errors
        };
    }
    
    /**
     * Secure API request with authentication
     * @param {string} url - API endpoint
     * @param {object} options - Fetch options
     * @returns {Promise} - Fetch promise
     */
    static secureRequest(url, options = {}) {
        const token = localStorage.getItem('auth_token');
        
        const defaultOptions = {
            headers: {
                'Content-Type': 'application/json',
                'X-Requested-With': 'XMLHttpRequest'
            },
            credentials: 'same-origin'
        };
        
        if (token) {
            defaultOptions.headers['Authorization'] = `Bearer ${token}`;
        }
        
        const finalOptions = {
            ...defaultOptions,
            ...options,
            headers: {
                ...defaultOptions.headers,
                ...options.headers
            }
        };
        
        return fetch(url, finalOptions);
    }
    
    /**
     * Log security events
     * @param {string} event - Event type
     * @param {object} details - Event details
     */
    static logSecurityEvent(event, details = {}) {
        const logEntry = {
            timestamp: new Date().toISOString(),
            event: event,
            details: details,
            userAgent: navigator.userAgent,
            url: window.location.href
        };
        
        console.log('[SECURITY]', logEntry);
        
        // In production, send to logging service
        // this.sendToLoggingService(logEntry);
    }
    
    /**
     * Check if user is authenticated
     * @returns {boolean} - True if authenticated
     */
    static isAuthenticated() {
        const token = localStorage.getItem('auth_token');
        const userType = localStorage.getItem('user_type');
        const userId = localStorage.getItem('user_id');
        
        return !!(token && userType && userId);
    }
    
    /**
     * Clear authentication data
     */
    static logout() {
        localStorage.removeItem('auth_token');
        localStorage.removeItem('user_type');
        localStorage.removeItem('user_id');
        this.logSecurityEvent('LOGOUT');
    }
    
    /**
     * Prevent XSS in innerHTML usage
     * @param {string} html - HTML string to sanitize
     * @returns {string} - Sanitized HTML
     */
    static sanitizeHTML(html) {
        const div = document.createElement('div');
        div.textContent = html;
        return div.innerHTML;
    }
}

// Export for use in other files
if (typeof module !== 'undefined' && module.exports) {
    module.exports = SecurityUtils;
} 